### PR TITLE
DAOS-5309 Data-Migration: Implementation of Server Addition.

### PR DESCRIPTION
**Pull request by PetFet**
This patch implements Server Addition as well as provides basic
tests for server addition.

- Changed testing setup so that it now supports variable
pool size and not just max number of ranks.
- Added add_simple tests based on other simple tests. It starts with 4
servers and then adds a 5th for all different tests.

Signed-off-by: Peter Fetros <peter.fetros@intel.com>
Signed-off-by: Byron Marohn <byron.marohn@intel.com>

### CODE REVIEW

File: src/tests/suite/daos_add_simple.c [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+	{"ADDITION6: addition records with multiple snapshots",
+	 addition_snap_update_recs, addition_small_sub_setup, test_teardown},
+	{"ADDITION7: addition punch/records with multiple snapshots",
+	 addition_snap_punch_recs, addition_small_sub_setup, test_teardown},
+	{"ADDITION8: addition keys with multiple snapshots",
+	 addition_snap_update_keys, addition_small_sub_setup, test_teardown},
+	{"ADDITION9: addition keys/punch with multiple snapshots",
+	 addition_snap_punch_keys, addition_small_sub_setup, test_teardown},
+};
+
```
Comment (wangdi1) [inline src/tests/suite/daos_add_simple.c:483]: These tests seems almost same as rebuild_simple() ?  Just curious, how do you make sure the data will be migrated during rank add.

File: src/rebuild/srv.c [hunk 2]
```diff
@@ -1457,36 +1463,51 @@ regenerate_task_internal(struct ds_pool *pool, struct pool_target *tgts,
...
+
+	rc = regenerate_task_of_type(pool, PO_COMP_ST_DOWN, RB_OP_FAIL);
 	if (rc != 0)
 		return rc;
 
-	/* get all up targets */
-	rc = pool_map_find_up_tgts(pool->sp_map, &tgts, &tgts_cnt);
-	if (rc != 0) {
-		D_ERROR("failed to create up tgt list rc %d\n", rc);
+	rc = regenerate_task_of_type(pool, PO_COMP_ST_DRAIN, RB_OP_DRAIN);
```
Comment (wangdi1) [inline src/rebuild/srv.c:1498]: Maybe just find pull out all targets with DOWN|DRAIN|UP states, then regenerate the tasks by their state.  So it does not need to call regenerate_task_of_type 3 times?

**Comment by byron-marohn** [comment]
Removed the incomplete unit tests from this PR and opened DAOS-5674 to track adding them back in (and making sure they actually cover addition)

This pull request just implements server addition, and should be safe to land as it doesn't change any other functionality, nor affect any tests that will begin failing as a result of landing this patch. 

Would still appreciate review & suggestions of course on just this cut-down version.

File: src/common/pool_map.c [hunk 3]
```diff
@@ -934,12 +934,10 @@ pool_map_finalise(struct pool_map *map)
  * Install a component tree to a pool map.
  *
  * \param map		[IN]	The pool map to be initialized.
- * \param activate	[IN]	Activate pool components.
```
Comment (byron-marohn) [inline src/common/pool_map.c:937]: This "activate" logic caused a huge number of problems for addition, as many different components would get the pool map update with the new ranks in the NEW state, then helpfully "activate" them, causing pool version desync. All of the addition testing was with this activation code removed - it doesn't seem to be important to being able to run the servers, as apparently components are set to UPIN elsewhere during startup.

File: src/pool/srv_util.c [hunk 4]
```diff
@@ -332,10 +356,17 @@ update_one_tgt(struct pool_map *map, struct pool_target *target,
...
 			D_DEBUG(DF_DSMS, "change target %u/%u to UPIN %p\n",
 				target->ta_comp.co_rank,
 				target->ta_comp.co_index, map);
-			target->ta_comp.co_status = PO_COMP_ST_UPIN;
+			/*
+			 * Need to update this target AND all of its parents
+			 * domains from NEW -> UPIN
+			 */
+			rc = pool_map_activate_new_target(map,
+					target->ta_comp.co_id);
```
[the remark this answers is not in the document]
Comment (byron-marohn) [reply src/pool/srv_util.c:368]: doesn't fit with 80 character line limit

File: src/rebuild/srv.c (hunk 2 above)
Comment (byron-marohn) [reply src/rebuild/srv.c:1499]: I think this logic will be improved later for DAOS-2843 (prioritize rebuild/extend/reint/drain)

**Review by NiuYawei — APPROVED** [review]

File: src/object/cli_obj.c [hunk 5]
```diff
@@ -814,7 +814,7 @@ obj_shard_tgts_query(struct dc_object *obj, uint32_t map_ver, uint32_t shard,
 		D_GOTO(out, rc);
 
 	shard_tgt->st_rank	= obj_shard->do_target_rank;
-	shard_tgt->st_shard	= shard,
+	shard_tgt->st_shard	= obj_shard->do_shard,
```
Comment (wangdi1) [inline src/object/cli_obj.c:817]: This will conflict with https://github.com/daos-stack/daos/pull/3428

**Review by wangdi1 — DISMISSED** [review]
There is actually a few changes has been put to another patch. But if this is urgent, I do not mind land this first.

**Comment by rpadma2** [comment]
Two tests are known issues:
CSUM_log : DAOS-5745
DaosCoreRebuild : DAOS-5610
Leap-15 are known issues on all PRs.

**Review by wangdi1 — APPROVED** [review]

**Comment by rpadma2** [comment]
Can this PR be landed? I believe Niu approved it before... Not sure whether he is on vacation.

**Comment by johannlombardi** [comment]
@rpadma2 Could you please rebase on the tip of master?

**Comment by rpadma2** [comment]
> @rpadma2 Could you please rebase on the tip of master?

Done.

**Review by NiuYawei — APPROVED** [review]

**Pull request merged**